kong/plugins/ldap-auth/asn1.lua

```diff
@@ -27,183 +53,169 @@ _M.BERCLASS = {
 
-    -- Octet String
-    self.decoder["04"] = function(self, encStr, elen, pos)
-      return bunpack(encStr, "A" .. elen, pos)
-    end
+  function asn1_get_object(der, start, stop)
+    start = start or 0
+    stop = stop or #der
+    assert(stop >= start, fmt("invalid offset '%s', start: %s", stop, start))
+    assert(stop <= #der, "stop offset must less than length")
 
-    -- Null
-    self.decoder["05"] = function(self, encStr, elen, pos)
-      return pos, false
-    end
+    local s_der = ffi_cast("const unsigned char *", der)
+    strpp[0] = s_der + start
 
-    -- Object Identifier
-    self.decoder["06"] = function(self, encStr, elen, pos)
-      return self:decodeOID(encStr, elen, pos)
+    local ret = C.ASN1_get_object(strpp, lenp, tagp, classp, stop - start + 1)
+    if band(ret, 0x80) == 0x80 then
+        error("der with error encoding", ret)
     end
 
-    -- Context specific tags
-    self.decoder["30"] = function(self, encStr, elen, pos)
-      return self:decodeSeq(encStr, elen, pos)
+    local constructed = false
+    if band(ret, 0x20) == 0x20 then
+        constructed = true
     end
-  end,
 
-  registerTagDecoders = function(self, tagDecoders)
-    self:registerBaseDecoders()
-
-    for k, v in pairs(tagDecoders) do
-      self.decoder[k] = v
-    end
-  end,
+    return tagp[0], classp[0], tonumber(lenp[0]), strpp[0] - s_der, constructed
+  end
+end
+_M.asn1_get_object = asn1_get_object
 
-  decodeLength = function(encStr, pos)
-    local elen
 
-    pos, elen = bunpack(encStr, "C", pos)
-    if elen > 128 then
-      elen = elen - 128
-      local elenCalc = 0
-      local elenNext
+--[[
+Encoded LDAP Result: https://ldap.com/ldapv3-wire-protocol-reference-ldap-result/
 
-      for i = 1, elen do
-        elenCalc = elenCalc * 256
-        pos, elenNext = bunpack(encStr, "C", pos)
-        elenCalc = elenCalc + elenNext
-      end
+30 0c -- Begin the LDAPMessage sequence
+   02 01 03 -- The message ID (integer value 3)
+   69 07 -- Begin the add response protocol op
+      0a 01 00 -- success result code (enumerated value 0)
+      04 00 -- No matched DN (0-byte octet string)
+      04 00 -- No diagnostic message (0-byte octet string)
+--]]
 
-      elen = elenCalc
-    end
+local function parse_ldap_result(der)
+  local p = ffi_cast("const unsigned char *", der)
+  cucharpp[0] = p
+  local _, _, len, offset = asn1_get_object(der)
 
-    return pos, elen
-  end,
+  -- message ID (integer)
+  local asn1_int = C.d2i_ASN1_INTEGER(nil, cucharpp, #der)
+  local id = C.ASN1_INTEGER_get(asn1_int)
 
-  decodeSeq = function(self, encStr, len, pos)
-    local seq = {}
-    local sPos = 1
-    local sStr
+  _, _, len, offset = asn1_get_object(der, offset + len)
+  local _, op_int = bunpack(der, "C", offset - 1)
 
-    pos, sStr = bunpack(encStr, "A" .. len, pos)
+  local op = _M.intToBER(op_int)
 
-    while (sPos < len) do
-      local newSeq
+  -- success result code
+  cucharpp[0] = p + offset
```

Comment:
need check for buffer overflow here (compare `p+offset < #der`

Reply:
`asn1_get_object` Lua func contains the offset checking, seems fine